android: App crashes when drag up entry in scroll container

### To Reproduce:
Steps to reproduce the behaviour:
1. Open Fyne demo
2. Open Text tab in Widgets tab
3. Try to scroll Lorem ipsum entry (it should not fit into the screen)
4. The app is closing

### Device (please complete the following information):
 - OS: Android
 - Version: 8.0
 - Go version: 1.14.2
 - Fyne version: develop 04d654b06855c53d2a7c18caafbcca2161baff73

@stuartmscott It could be reproduced on desktop with `-tags mobile` too, but harder
Here is [video](https://imgur.com/a/FgKQ8oS)

Stacktrace
```
panic: runtime error: index out of range [17] with length 7

goroutine 38 [running]:
fyne.io/fyne/widget.(*entryRenderer).buildSelection(0xc0000c43c0)
        /home/max/Projects/fyne/widget/entry.go:1079 +0x63c
fyne.io/fyne/widget.(*entryRenderer).moveCursor(0xc0000c43c0)
        /home/max/Projects/fyne/widget/entry.go:1092 +0xc0
fyne.io/fyne/widget.(*entryRenderer).Refresh(0xc0000c43c0)
        /home/max/Projects/fyne/widget/entry.go:984 +0x1bb
fyne.io/fyne/widget.(*BaseWidget).Refresh(0xc0000ce900)
        /home/max/Projects/fyne/widget/widget.go:107 +0x51
fyne.io/fyne/widget.(*Entry).updateMousePointer(0xc0000ce900, 0xc0013f2b10, 0x0)
        /home/max/Projects/fyne/widget/entry.go:879 +0x113
fyne.io/fyne/widget.(*Entry).Dragged(0xc0000ce900, 0xc0013f2b10)
        /home/max/Projects/fyne/widget/entry.go:161 +0x43
created by fyne.io/fyne/internal/driver/gomobile.(*mobileDriver).tapMoveCanvas.func1
        /home/max/Projects/fyne/internal/driver/gomobile/driver.go:229 +0x4d
exit status 2
```

--- a/internal/driver/gomobile/canvas.go
+++ b/internal/driver/gomobile/canvas.go
@@ -330,6 +330,7 @@ func (c *mobileCanvas) tapUp(pos fyne.Position, tapID int,
 		c.dragging.DragEnd()
 
 		c.dragging = nil
+		return
 	}
 
 	duration := time.Since(c.lastTapDown[tapID])

--- a/internal/driver/gomobile/canvas.go
+++ b/internal/driver/gomobile/canvas.go
@@ -313,7 +313,6 @@ func (c *mobileCanvas) tapMove(pos fyne.Position, tapID int,
 			return
 		}
 	}
-	objPos = pos.Subtract(c.dragging.(fyne.CanvasObject).Position())
 
 	ev := new(fyne.DragEvent)
 	ev.Position = objPos

--- a/widget/entry.go
+++ b/widget/entry.go
@@ -165,7 +165,11 @@ func (e *Entry) DragEnd() {
 // It updates the selection accordingly.
 // Implements: fyne.Draggable
 func (e *Entry) Dragged(d *fyne.DragEvent) {
-	e.selecting = true
+	if !e.selecting {
+		e.selectRow, e.selectColumn = e.getRowCol(&d.PointEvent)
+
+		e.selecting = true
+	}
 	e.updateMousePointer(&d.PointEvent, false)
 }
 
@@ -338,6 +342,10 @@ func (e *Entry) SetText(text string) {
 // Tapped is called when this entry has been tapped so we should update the cursor position.
 // Implements: fyne.Tappable
 func (e *Entry) Tapped(ev *fyne.PointEvent) {
+	if fyne.CurrentDevice().IsMobile() && e.selecting {
+		e.selecting = false
+		return
+	}
 	e.updateMousePointer(ev, false)
 }
 
@@ -871,11 +879,7 @@ func (e *Entry) textWrap() fyne.TextWrap {
 	return e.Wrapping
 }
 
-func (e *Entry) updateMousePointer(ev *fyne.PointEvent, rightClick bool) {
-	if !e.focused && !e.Disabled() {
-		e.FocusGained()
-	}
-
+func (e *Entry) getRowCol(ev *fyne.PointEvent) (int, int) {
 	rowHeight := e.textProvider().charMinSize().Height
 	row := int(math.Floor(float64(ev.Position.Y-theme.Padding()) / float64(rowHeight)))
 	col := 0
@@ -887,6 +891,16 @@ func (e *Entry) updateMousePointer(ev *fyne.PointEvent, rightClick bool) {
 	} else {
 		col = e.cursorColAt(e.textProvider().row(row), ev.Position)
 	}
+
+	return row, col
+}
+
+func (e *Entry) updateMousePointer(ev *fyne.PointEvent, rightClick bool) {
+	if !e.focused && !e.Disabled() {
+		e.FocusGained()
+	}
+
+	row, col := e.getRowCol(ev)
 
 	e.Lock()
 	if !rightClick || rightClick && !e.selecting {

--- a/widget/entry.go
+++ b/widget/entry.go
@@ -344,7 +344,6 @@ func (e *Entry) SetText(text string) {
 func (e *Entry) Tapped(ev *fyne.PointEvent) {
 	if fyne.CurrentDevice().IsMobile() && e.selecting {
 		e.selecting = false
-		return
 	}
 	e.updateMousePointer(ev, false)
 }

--- a/widget/entry.go
+++ b/widget/entry.go
@@ -1086,6 +1086,8 @@ func (r *entryRenderer) buildSelection() {
 		r.selection = r.selection[:rowCount]
 	}
 
+	r.entry.Lock()
+	defer r.entry.Unlock()
 	// build a rectangle for each row and add it to r.selection
 	for i := 0; i < rowCount; i++ {
 		if len(r.selection) <= i {

